src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -193,37 +186,98 @@ public async Task UpdatePrimaryBranchSolutionAsync(AssetProvider assetProvider,
             Checksum solutionChecksum,
             bool fromPrimaryBranch,
             int workspaceVersion,
+            ProjectId? projectId,
             CancellationToken cancellationToken)
+        {
+            try
+            {
+                return projectId == null
+                    ? await GetFullSolutionAsync(assetProvider, solutionChecksum, fromPrimaryBranch, workspaceVersion, cancellationToken).ConfigureAwait(false)
+                    : await GetProjectSubsetSolutionAsync(assetProvider, solutionChecksum, projectId, cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception e) when (FatalError.ReportAndPropagateUnlessCanceled(e, cancellationToken))
+            {
+                throw ExceptionUtilities.Unreachable;
+            }
+        }
+
+        private async ValueTask<Solution> GetFullSolutionAsync(AssetProvider assetProvider, Checksum solutionChecksum, bool fromPrimaryBranch, int workspaceVersion, CancellationToken cancellationToken)
```

Comment:
>GetFullSolutionAsync [](http://example.com/codeflow?start=42&length=20)

nit: move this method closer to CreateFullSolution_NoLockAsync

Reply:
will do.